src/providers/wfs/qgswfsprovider.cpp

```diff
@@ -852,6 +852,16 @@ bool QgsWFSProvider::addFeatures( QgsFeatureList &flist )
   {
     //transaction successful. Add the features to the cache
     QStringList idList = insertedFeatureIds( serverResponse );
+    /* Fix issue with GeoServer and shapefile feature stores when no real
+       feature id are returned but new0 returned instead of the featureId*/
+  for ( const QString &v : idList )
```

Comment:
@nyalldawson  This what I read before thinking that in this particular case a range loop would not have triggered a deep copy, but I might have misunderstood that https://www.dvratil.cz/2015/06/qt-containers-and-c11-range-based-loops/
BTW: no problems to change into a Q_FOREACH